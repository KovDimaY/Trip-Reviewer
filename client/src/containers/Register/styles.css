.register-container {
    text-align: center;
}

.register-container .form_element {
    margin-top: 20px;
}

.register-container h2 {
    font-size: 33px;
    font-weight: 300;
    color: #02bcd4;
}

.register-container input {
    font-size: 20px;
    border-radius: 3px;
    border: 1px solid #ebebeb;
    box-sizing: border-box;
    padding: 11px 31px;
    background: #f7f7f7;
    font-weight: 300;
    width: 260px;
}

.register-container button {
    font-size: 20px;
    padding: 10px 52px;
    background: #02bcd4;
    color: #fff;
    font-weight: 300;
    border-radius: 3px;
    border: none;
    margin-top: 20px;
    width: 260px;
}

.register-container .error {
    color: #FF5722;
    margin-top: 10px;
    font-weight: 600;
}

.register-container .current_users {
    width: 90%;
    max-width: 500px;
    margin: auto;
    margin-top: 50px;
}

.register-container .current_users h3 {
    font-size: 24px;
    font-weight: 300;
    color: orange;
}

.register-container .current_users th, td {
    text-align: center;
}

.register-container .current_users tr:nth-child(even) {
    background-color: #f2f2f2;
}