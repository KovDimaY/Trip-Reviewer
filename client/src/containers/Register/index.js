import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import { getUsers, userRegister } from '../../actions';
import { USER_PROFILE } from '../../constants/routes';

import './styles.css';

class Register extends PureComponent {
    state = {
      formData: {
        name: '',
        lastname: '',
        email: '',
        password: '',
      },
      hideError: {
        name: false,
        lastname: false,
        email: false,
        password: false,
      },
    };

    componentWillMount() {
      this.props.dispatch(getUsers());
    }

    componentWillReceiveProps(nextProps) {
      const { login } = nextProps.users;
      if (login.isAuth) {
        this.props.history.push(USER_PROFILE);
      } else {
        this.setState({
          hideError: {
            name: false,
            lastname: false,
            email: false,
            password: false,
          },
        });
      }
    }

    getErrorClass(fieldName) {
      return this.formFieldHasError(fieldName) ? 'field-error' : '';
    }

    handleInput = (event) => {
      const newFormData = { ...this.state.formData };
      const newHideError = { ...this.state.hideError };
      const { value, name } = event.target;

      newFormData[name] = value;
      newHideError[name] = true;

      this.setState({
        formData: newFormData,
        hideError: newHideError,
      });
    }

    submitForm = (event) => {
      event.preventDefault();

      this.props.dispatch(userRegister(this.state.formData));
    }

    formFieldHasError(fieldName) {
      const { register } = this.props.users;
      const { hideError } = this.state;
      const errors = register && register.error && register.error.errors;

      return errors && errors[fieldName] && !hideError[fieldName];
    }

    renderError(fieldName) {
      if (this.formFieldHasError(fieldName)) {
        const error = this.props.users.register.error.errors[fieldName];

        return (
          <div className="error">
            {error.message}
          </div>
        );
      }
      return null;
    }

    renderUsersTable = ({ users }) => (
      users
        ? users.map(item => (
          <tr key={item._id}>
            <td>
              {item.name}
            </td>
            <td>
              {item.lastname}
            </td>
          </tr>
        ))
        : null
    )

    render() {
      const { users } = this.props;
      const {
        name, lastname,
        email, password,
      } = this.state.formData;

      return (
        <div className="register-container">
          <form onSubmit={this.submitForm}>
            <h2>
              Sing Up
            </h2>

            <div className="form_element">
              <input
                type="text"
                className={this.getErrorClass('name')}
                placeholder="Enter name"
                value={name}
                name="name"
                onChange={this.handleInput}
              />
            </div>
            { this.renderError('name') }

            <div className="form_element">
              <input
                type="text"
                className={this.getErrorClass('lastname')}
                placeholder="Enter Lastname"
                value={lastname}
                name="lastname"
                onChange={this.handleInput}
              />
            </div>
            { this.renderError('lastname') }

            <div className="form_element">
              <input
                type="email"
                className={this.getErrorClass('email')}
                placeholder="Enter Email"
                value={email}
                name="email"
                onChange={this.handleInput}
              />
            </div>
            { this.renderError('email') }

            <div className="form_element">
              <input
                type="password"
                className={this.getErrorClass('password')}
                placeholder="Enter Password"
                value={password}
                name="password"
                onChange={this.handleInput}
              />
            </div>
            { this.renderError('password') }

            <button type="submit">
              Sign Up
            </button>

          </form>
          <div className="current_users">
            <h3>
              Current users:
            </h3>
            <table>
              <thead>
                <tr>
                  <th>
                    Name
                  </th>
                  <th>
                    Lastname
                  </th>
                </tr>
              </thead>
              <tbody>
                {this.renderUsersTable(users)}
              </tbody>
            </table>
          </div>
        </div>
      );
    }
}

Register.propTypes = {
  users: PropTypes.object.isRequired,
  history: PropTypes.object.isRequired,
  dispatch: PropTypes.func.isRequired,
};

function mapStateToProps(state) {
  return {
    users: state.users,
  };
}

export default connect(mapStateToProps)(Register);
