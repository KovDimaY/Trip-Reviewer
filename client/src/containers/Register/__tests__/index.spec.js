import React from 'react';
import { create } from 'react-test-renderer';
 
import Register from './../../Register';
import Root from './../../../hoc/root';

const mockComponent = props => {
    return (
        <Root>
            <Register {...props} />
        </Root>
    );
};
 
describe('<Register />', () => {
    it('should render component', () => {
        const props = {
            user: {}
        };
        const tree = create(mockComponent(props)).toJSON();
        
        expect(tree).toMatchSnapshot();
    });
});