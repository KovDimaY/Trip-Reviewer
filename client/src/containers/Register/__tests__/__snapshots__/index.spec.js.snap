// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Register /> should render component 1`] = `
<div
  className="rl_container"
>
  <form
    onSubmit={[Function]}
  >
    <h2>
      Sing Up
    </h2>
    <div
      className="form_element"
    >
      <input
        onChange={[Function]}
        placeholder="Enter name"
        type="text"
        value=""
      />
    </div>
    <div
      className="form_element"
    >
      <input
        onChange={[Function]}
        placeholder="Enter Lastname"
        type="text"
        value=""
      />
    </div>
    <div
      className="form_element"
    >
      <input
        onChange={[Function]}
        placeholder="Enter Email"
        type="email"
        value=""
      />
    </div>
    <div
      className="form_element"
    >
      <input
        onChange={[Function]}
        placeholder="Enter Password"
        type="password"
        value=""
      />
    </div>
    <button
      type="submit"
    >
      Sign Up
    </button>
    <div
      className="error"
    >
      
    </div>
  </form>
  <div
    className="current_users"
  >
    <h4>
      Current users:
    </h4>
    <table>
      <thead>
        <tr>
          <th>
            Name
          </th>
          <th>
            Lastname
          </th>
        </tr>
      </thead>
      <tbody />
    </table>
  </div>
</div>
`;
